pkg/certs/certs.go

```diff
@@ -121,13 +117,19 @@ func (c *Certs) watchSymlinks() (err error) {
 // and the old (if any) certificates and keys will continue
 // to be used.
 func (c *Certs) watch() (err error) {
+	c.Lock()
+	defer c.Unlock()
+	c.cert, c.caCerts, err = c.loadCert(c.certFile, c.keyFile)
```

Comment:
this pattern is wrong @aead the previous code is correct, we would have loaded the certs before watching already. So in the flow again doing this is not needed... unless file system notified of a change.

Reply:
Why, the `watchSymlinks` does it exactly like that (even before this change)- what is the difference between `watch` and `watchSymlinks`?

Reply:
if you want to cleanup I would suggest you do it a bit differently, move the loadCert into the New() function itself @aead here is the diff

```diff
diff --git a/pkg/certs/certs.go b/pkg/certs/certs.go
index c91a07ac1..d45799242 100644
--- a/pkg/certs/certs.go
+++ b/pkg/certs/certs.go
@@ -60,12 +60,20 @@ func New(certFile, keyFile string, loadCert LoadX509KeyPairFunc) (*Certs, error)
                e: make(chan notify.EventInfo, 1),
        }
 
+       // Load certs if possible, we don't need to proceed further
+       // if certs cannot be loaded to start with.
+       var err error
+       c.cert, c.caCerts, err = c.loadCert(c.certFile, c.keyFile)
+       if err != nil {
+               return nil, err
+       }
+
        if checkSymlink(certFile) && checkSymlink(keyFile) {
-               if err := c.watchSymlinks(); err != nil {
+               if err = c.watchSymlinks(); err != nil {
                        return nil, err
                }
        } else {
-               if err := c.watch(); err != nil {
+               if err = c.watch(); err != nil {
                        return nil, err
                }
        }
@@ -84,12 +92,6 @@ func checkSymlink(file string) bool {
 // watchSymlinks reloads symlinked files since fsnotify cannot watch
 // on symbolic links.
 func (c *Certs) watchSymlinks() (err error) {
-       c.Lock()
-       defer c.Unlock()
-       c.cert, c.caCerts, err = c.loadCert(c.certFile, c.keyFile)
-       if err != nil {
-               return err
-       }
        go func() {
                for {
                        select {
@@ -117,13 +119,6 @@ func (c *Certs) watchSymlinks() (err error) {
 // and the old (if any) certificates and keys will continue
 // to be used.
 func (c *Certs) watch() (err error) {
-       c.Lock()
-       defer c.Unlock()
-       c.cert, c.caCerts, err = c.loadCert(c.certFile, c.keyFile)
-       if err != nil {
-               return err
-       }
-
        defer func() {
                if err != nil {
                        // Stop any watches previously setup after an error.
```

This is more cleaner and readable.

Reply:
Agree - yeah, done